packages/job-worker/src/playout/playout.ts

```diff
@@ -837,140 +795,140 @@ export async function onPartPlaybackStarted(context: JobContext, data: OnPartPla
-							)
-						} else if (!previousPartInstance.timings?.duration) {
-							onPartHasStoppedPlaying(cache, previousPartInstance, data.startedPlayback)
-						}
-					}
+			await _onPartPlaybackStarted(context, cache, data)
+		}
+	)
+}
+async function _onPartPlaybackStarted(context: JobContext, cache: CacheForPlayout, data: OnPartPlaybackStartedProps) {
+	const playingPartInstance = cache.PartInstances.findOne(data.partInstanceId)
+	if (!playingPartInstance)
+		throw new Error(`PartInstance "${data.partInstanceId}" in RundownPlayst "${data.playlistId}" not found!`)
+
+	// make sure we don't run multiple times, even if TSR calls us multiple times
+	const isPlaying = playingPartInstance.timings?.startedPlayback && !playingPartInstance.timings?.stoppedPlayback
+	if (!isPlaying) {
+		logger.debug(
+			`Playout reports PartInstance "${data.partInstanceId}" has started playback on timestamp ${new Date(
+				data.startedPlayback
+			).toISOString()}`
+		)
 
-					reportPartInstanceHasStarted(context, cache, playingPartInstance, data.startedPlayback)
-				} else if (playlist.nextPartInstanceId === data.partInstanceId) {
-					// this is the next part, clearly an autoNext has taken place
-					if (playlist.currentPartInstanceId) {
-						if (!currentPartInstance) {
-							// We couldn't find the previous part: this is not a critical issue, but is clearly is a symptom of a larger issue
-							logger.error(
-								`Previous PartInstance "${playlist.currentPartInstanceId}" on RundownPlaylist "${playlist._id}" could not be found.`
-							)
-						} else if (!currentPartInstance.timings?.duration) {
-							onPartHasStoppedPlaying(cache, currentPartInstance, data.startedPlayback)
-						}
-					}
+		const playlist = cache.Playlist.doc
 
-					cache.Playlist.update({
-						$set: {
-							previousPartInstanceId: playlist.currentPartInstanceId,
-							currentPartInstanceId: playingPartInstance._id,
-							holdState: RundownHoldState.NONE,
-						},
-					})
+		const rundown = cache.Rundowns.findOne(playingPartInstance.rundownId)
+		if (!rundown) throw new Error(`Rundown "${playingPartInstance.rundownId}" not found!`)
 
-					reportPartInstanceHasStarted(context, cache, playingPartInstance, data.startedPlayback)
+		const { currentPartInstance, previousPartInstance } = getSelectedPartInstancesFromCache(cache)
 
-					// Update generated properties on the newly playing partInstance
-					const currentRundown = currentPartInstance
-						? cache.Rundowns.findOne(currentPartInstance.rundownId)
-						: undefined
-					const showStyleRundown = currentRundown ?? rundown
-					const showStyle = await context.getShowStyleCompound(
-						showStyleRundown.showStyleVariantId,
-						showStyleRundown.showStyleBaseId
+		if (playlist.currentPartInstanceId === data.partInstanceId) {
+			// this is the current part, it has just started playback
+			if (playlist.previousPartInstanceId) {
+				if (!previousPartInstance) {
```

Comment:
You are four layers of nesting deep here... :( 
A good rule of thumb is to try and only have one layer of nesting for a method

Reply:
ref: https://github.com/nrkno/sofie-core/pull/737#discussion_r912742322